### PR TITLE
Implement rulesets

### PR DESCRIPTION
**Pull request by liaodaniel**
This PR includes both the organization and repository rulesets API.

Resolves https://github.com/google/go-github/issues/2789

Signed-off-by: Daniel Liao [10663736+liaodaniel@users.noreply.github.com](mailto:10663736+liaodaniel@users.noreply.github.com)

### CODE REVIEW

**Comment by liaodaniel** [comment]
Linter appears to be failing due to https://github.com/golangci/golangci-lint/issues/3862. Guess I'll wait until they make a fix for it

File: github/orgs_rules.go
Comment (gmlewis) [inline github/orgs_rules.go:64]:
```suggestion
	var ruleset *Ruleset
```

File: github/orgs_rules.go
Comment (gmlewis) [inline github/orgs_rules.go:84]:
```suggestion
	var ruleset *Ruleset
```

File: github/repos_rules.go
Comment (gmlewis) [inline github/repos_rules.go:16]:
```suggestion
	ActorID *int64 `json:"actor_id,omitempty"`
```

File: github/repos_rules.go
Comment (gmlewis) [inline github/repos_rules.go:18]:
```suggestion
	ActorType *string `json:"actor_type,omitempty"`
```

File: github/repos_rules.go
Comment (gmlewis) [inline github/repos_rules.go:34]:
```suggestion
	Include []string `json:"include,omitempty"`
	Exclude []string `json:"exclude,omitempty"`
```

File: github/repos_rules.go
Comment (gmlewis) [inline github/repos_rules.go:40]: It is challenging to tell if fields need to be `omitempty` from the docs, but it seems like `include` and `exclude` would be a good candidate for optionally including one or the other.

```suggestion
	Include   []string `json:"include,omitempty"`
	Exclude   []string `json:"exclude,omitempty"`
```

File: github/repos_rules.go
Comment (gmlewis) [inline github/repos_rules.go:94]: Instead of `interface{}`, I'm wondering if this would be better to have a `*json.RawMessage` here like we do in other places in this repo?

File: github/repos_rules.go
Comment (gmlewis) [inline github/repos_rules.go:114]: For these coverage warnings, it would be nice if you could add some tests that placed bogus JSON in the data such that these `json.Unmarshal` lines flagged by codecov would fail.  Something like "{]" usually does the trick.

File: github/repos_rules.go
Comment (gmlewis) [inline github/repos_rules.go:196]: There's no need to break this line here... please add the return value here to the end of line 195.

File: github/repos_rules.go
Comment (gmlewis) [inline github/repos_rules.go:271]: Is `*[]BypassActor` necessary here?  Can't it be `[]*BypassActor` instead, which is preferred?

File: github/repos_rules.go
Comment (gmlewis) [inline github/repos_rules.go:275]: Is `*[]RepositoryRule` necessary here?  Can't it be `[]*RepositoryRule` instead, which is preferred?

File: github/repos_rules.go
Comment (gmlewis) [inline github/repos_rules.go:299]:
```suggestion
// If includesParents is true, rulesets configured at the organization level that apply to the repository will be returned.
```

**Review by gmlewis — CHANGES_REQUESTED** [review]
Thank you, @liaodaniel !
This looks fantastic!  I have just a few change requests, and feel free to disagree on any of them. As I commented, it is challenging to tell if we have the right combination of `omitempty`, but this looks good to me if I haven't made comments otherwise.

File: github/repos_rules.go
Comment (liaodaniel) [reply github/repos_rules.go:94]: Can I get some guidance on how this might be done? I took a look at some of the other examples and they seem to suggest to allow users to pass in `*json.RawMessage` types as the parameters? I am on the fence with this one as I feel like you lose out on all the typed constructs when you switch over to this it in this particular case?

```
# Example of creating a pull request rule
func NewPullRequestRule(params *json.RawMessage) (rule *RepositoryRule) {
	return &RepositoryRule{
		Type:       "pull_request",
		Parameters: params,
	}
}

pullRequestParams := json.RawMessage(`{
	"dismiss_stale_reviews_on_push": true,
	"require_code_owner_review": true,
	"require_last_push_approval": true,
	"required_approving_review_count": 1,
	"required_review_thread_resolution": true
}`)

rule := NewPullRequestRule(&pullRequestParams),
```

Comment (gmlewis) [reply github/repos_rules.go:94]: Sorry, I'm OOO today and will take a look tomorrow.

Comment (gmlewis) [reply github/repos_rules.go:94]: In general, you don't have to give up strong typing with `json.RawMessage`... you can always marshal other strongly-typed structs into the raw message.  It is just another way to handle variadic sub-messages within a JSON blob.
So I'm not sure you lose any functionality with `json.RawMessage` over the use of `any`... I think it is just a cleaner way of implementing the interface... but maybe I'm not understanding your question.

Comment (liaodaniel) [reply github/repos_rules.go]: Ok, thank you. Think I may have interpreted what you meant incorrectly previously.

Was this what you had in mind? See https://github.com/google/go-github/pull/2795/commits/5618f6f4a68cd1cea4f355cda1dca67367ac50d3

**Review by gmlewis — APPROVED** [review]
Thank you, @liaodaniel !
LGTM.

Awaiting second LGTM+Approval from any other contributor to this repo before merging.

**Review by valbeat — APPROVED** [review]
LGTM!

**Comment by gmlewis** [comment]
Thank you, @valbeat !
Merging.

**Pull request merged**
